sdk/batch/azure-batch/tests/decorators.py

```diff
@@ -0,0 +1,153 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+import urllib.parse as url_parse
+
+from azure.core.exceptions import ResourceNotFoundError
+from azure.core.pipeline.policies import ContentDecodePolicy
+from azure.core.pipeline.transport import AioHttpTransport, RequestsTransport
+
+from devtools_testutils import trim_kwargs_from_test_function
+from devtools_testutils.helpers import is_live_and_not_recording
+from devtools_testutils.proxy_testcase import (
+    get_test_id,
+    start_record_or_playback,
+    transform_request,
+    stop_record_or_playback,
+)
+
+from async_wrapper import async_wrapper
+
+
+# A modified version of devtools_testutils.aio.recorded_by_proxy_async
+# that modifies AioHttpTransport.send and RequestsTransport.send to make
+# both async and sync calls work with the proxy.
+def recorded_by_proxy_async(test_func):
```

Comment:
It would be great if we could use the shared eng sys tooling in this repo so you automatically get updates and bug fixes. Maybe I'm missing something, but what's the reason for needing to have sync/async under this single decorator?

Reply:
We are running both the aio/async and sync test using the same suite of test code in the async style by wrapping the sync function call to return coroutine, so that we do need to duplicate our test code for sync/ async clients. To achieve this, we need a decorated that works with both async and sync client, so we modified `devtools_testutils.aio.recorded_by_proxy_async` a little bit to meet our need.

Reply:
When we migrating from Track1, we already had dozens of tests which covered a lot of scenarios for the sync client. Then we found there is an async client in Track2, and would like it to have same tests and coverage as sync client. To avoid duplicating test code or maintaining two suite of identical test code, we figured out we could use the same test code to test both the sync and async test with a wrapper/ adaptor, like what we do in the Track2 C# SDK.

Reply:
I believe customers won't use the sync client in the async way like what we wrote in tests. The test is in the async style and can test the async client without issues, I think we won't miss out bugs in the async side.

Reply:
if the tests dont reflect customer usage, how is it being tested ? Outside of samples, our customers tend to look at tests as well for ideas/best practices on how to use libraries.

Reply:
The wrapper function just act like a helper function to make the result better for assertion, we did cover both sync/ async client. We wrote tests to make sure the client is working as expected, basically it will put up the right params for the API call and deserialize the response correctly.

Reply:
Will python consider integrating the `async-sync` test into the testing framework like C# do?: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/README.md#sync-async-tests. That solves our problem as well.

Reply:
Planning on re-visiting this later. We might need to address this in another PR.